Extend JSON support in the CLI

slimhazard opened the pull request:
Adds -j forms for:
* status
* vcl.list
* param.show
* ban.list
* storage.list
* panic.show

And updates docs.

bsdphk commented:
Could you provide examples of the produced JSON in a comment here ?  That will make bugwashing easier on monday.

slimhazard commented:
``status -j`` output (from b00004.log):
```
[ 2, ["status", "-j"], 1538031732.632, "running"
]

[ 2, ["status", "-j"], 1538031733.764, "stopped"
]
```

slimhazard commented:
```
vcl.list -j

[ 2, ["vcl.list", "-j"], 1538031760.660,
  {
    "status": "available",
    "state": "auto",
    "temperature": "warm",
    "name": "vcl1"
},
  {
    "status": "active",
    "state": "auto",
    "temperature": "warm",
    "name": "vcl2"
}
]

[ 2, ["vcl.list", "-j"], 1538031760.816,
  {
    "status": "available",
    "state": "auto",
    "temperature": "warm",
    "busy": 0,
    "name": "vcl1"
},
  {
    "status": "active",
    "state": "auto",
    "temperature": "warm",
    "busy": 0,
    "name": "vcl2"
}
]

[ 2, ["vcl.list", "-j"], 1538031961.544,
  {
    "status": "active",
    "state": "auto",
    "temperature": "warm",
    "busy": 1,
    "name": "vcl1"
},
  {
    "status": "available",
    "state": "auto",
    "temperature": "warm",
    "busy": 1,
    "name": "vcl2",
    "labels": 2
},
  {
    "status": "available",
    "state": "label",
    "temperature": "warm",
    "busy": 0,
    "name": "foo",
    "label": {
      "name": "vcl2"
      }
},
  {
    "status": "available",
    "state": "label",
    "temperature": "warm",
    "busy": 0,
    "name": "bar",
    "label": {
      "name": "vcl2"
      }
}
]

[ 2, ["vcl.list", "-j"], 1538031965.484,
  {
    "status": "active",
    "state": "warm",
    "temperature": "warm",
    "name": "vcl1",
    "labels": 3
},
  {
    "status": "available",
    "state": "auto",
    "temperature": "warm",
    "name": "vcl2"
},
  {
    "status": "available",
    "state": "label",
    "temperature": "warm",
    "name": "label1",
    "label": {
      "name": "vcl1"
    }
},
  {
    "status": "available",
    "state": "label",
    "temperature": "warm",
    "name": "label2",
    "label": {
      "name": "vcl1"
    }
},
  {
    "status": "available",
    "state": "label",
    "temperature": "warm",
    "name": "label3",
    "label": {
      "name": "vcl1"
    }
}
]

[ 2, ["vcl.list", "-j"], 1538031965.912,
  {
    "status": "active",
    "state": "warm",
    "temperature": "warm",
    "busy": 1,
    "name": "vcl1",
    "labels": 4
},
  {
    "status": "available",
    "state": "auto",
    "temperature": "warm",
    "busy": 0,
    "name": "vcl2"
},
  {
    "status": "available",
    "state": "label",
    "temperature": "warm",
    "busy": 0,
    "name": "label1",
    "label": {
      "name": "vcl1"
      }
},
  {
    "status": "available",
    "state": "label",
    "temperature": "warm",
    "busy": 0,
    "name": "label2",
    "label": {
      "name": "vcl1"
      }
},
  {
    "status": "available",
    "state": "label",
    "temperature": "warm",
    "busy": 0,
    "name": "label3",
    "label": {
      "name": "vcl1"
      }
},
  {
    "status": "available",
    "state": "label",
    "temperature": "warm",
    "busy": 0,
    "name": "snarf",
    "label": {
      "name": "vcl1"
      }
}
]
```

slimhazard commented:
```
param.show -j pool_req
[ 2, ["param.show", "-j", "pool_req"], 1538031772.064,
  {
    "name": "pool_req",
    "implemented": true,
    "value": {
        "min_pool": 10,
        "max_pool": 100,
        "max_age": 10
    },
    "default": "10,100,10",
    "description": "Parameters for per worker pool request memory pool.\nThe three numbers are:\n\tmin_pool\tminimum size of free pool.\n\tmax_pool\tmaximum size of free pool.\n\tmax_age\tmax age of free element.",
    "flags": [
    
]
}
]

param.show -j pool_sess
[ 2, ["param.show", "-j", "pool_sess"], 1538031772.108,
  {
    "name": "pool_sess",
    "implemented": true,
    "value": {
        "min_pool": 10,
        "max_pool": 100,
        "max_age": 10
    },
    "default": "10,100,10",
    "description": "Parameters for per worker pool session memory pool.\nThe three numbers are:\n\tmin_pool\tminimum size of free pool.\n\tmax_pool\tmaximum size of free pool.\n\tmax_age\tmax age of free element.",
    "flags": [
    
]
}
]

param.show -j changed

[ 2, ["param.show", "-j", "changed"], 1538031772.152,
  {
    "name": "auto_restart",
    "implemented": true,
    "value": false,
    "units": "bool",
    "default": "on",
    "description": "Automatically restart the child/worker process if it dies.",
    "flags": [
    
]
},
  {
    "name": "debug",
    "implemented": true,
    "value": "+vtc_mode",
    "default": "none",
    "description": "Enable/Disable various kinds of debugging.\n\tnone\tDisable all debugging\n\nUse +/- prefix to set/reset individual bits:\n\treq_state\tVSL Request state engine\n\tworkspace\tVSL Workspace operations\n\twaiter\tVSL Waiter internals\n\twaitinglist\tVSL Waitinglist events\n\tsyncvsl\tMake VSL synchronous\n\thashedge\tEdge cases in Hash\n\tvclrel\tRapid VCL release\n\tlurker\tVSL Ban lurker\n\tesi_chop\tChop ESI fetch to bits\n\tflush_head\tFlush after http1 head\n\tvtc_mode\tVarnishtest Mode\n\twitness\tEmit WITNESS lock records\n\tvsm_keep\tKeep the VSM file on restart\n\tdrop_pools\tDrop thread pools (testing)\n\tslow_acceptor\tSlow down Acceptor\n\th2_nocheck\tDisable various H2 checks\n\tvmod_so_keep\tKeep copied VMOD libraries\n\tprocessors\tFetch/Deliver processors\n\tprotocol\tProtocol debugging",
    "flags": [
    
]
},
  {
    "name": "prefer_ipv6",
    "implemented": true,
    "value": true,
    "units": "bool",
    "default": "off",
    "description": "Prefer IPv6 address when connecting to backends which have both IPv4 and IPv6 addresses.",
    "flags": [
    
]
},
  {
    "name": "syslog_cli_traffic",
    "implemented": true,
    "value": false,
    "units": "bool",
    "default": "on",
    "description": "Log all CLI traffic to syslog(LOG_INFO).",
    "flags": [
    
]
},
  {
    "name": "thread_pool_max",
    "implemented": true,
    "value": 4294967295,
    "units": "threads",
    "default": "5000",
    "minimum": "51",
    "description": "The maximum number of worker threads in each pool. The minimum value depends on thread_pool_min.\n\nDo not set this higher than you have to, since excess worker threads soak up RAM and CPU and generally just get in the way of getting work done.",
    "flags": [
      "delayed_effect"
]
},
  {
    "name": "thread_pool_min",
    "implemented": true,
    "value": 51,
    "units": "threads",
    "default": "100",
    "maximum": "4294967295",
    "description": "The minimum number of worker threads in each pool. The maximum value depends on thread_pool_max.\n\nIncreasing this may help ramp up faster from low load situations or when threads have expired.\n\nMinimum is 10 threads.",
    "flags": [
      "delayed_effect"
]
},
  {
    "name": "vmod_dir",
    "implemented": true,
    "value": "/home/geoff/Develop/varnish/varnish-cache/lib/libvmod_std/.libs:/home/geoff/Develop/varnish/varnish-cache/lib/libvmod_debug/.libs:/home/geoff/Develop/varnish/varnish-cache/lib/libvmod_directors/.libs:/home/geoff/Develop/varnish/varnish-cache/lib/libvmod_purge/.libs:/home/geoff/Develop/varnish/varnish-cache/lib/libvmod_vtc/.libs:/home/geoff/Develop/varnish/varnish-cache/lib/libvmod_blob/.libs:/home/geoff/Develop/varnish/varnish-cache/lib/libvmod_unix/.libs:/home/geoff/Develop/varnish/varnish-cache/lib/libvmod_proxy/.libs",
    "default": "/usr/local/lib/varnish/vmods",
    "description": "Old name for vmod_path, use that instead.",
    "flags": [
    
]
},
  {
    "name": "vmod_path",
    "implemented": true,
    "value": "/home/geoff/Develop/varnish/varnish-cache/lib/libvmod_std/.libs:/home/geoff/Develop/varnish/varnish-cache/lib/libvmod_debug/.libs:/home/geoff/Develop/varnish/varnish-cache/lib/libvmod_directors/.libs:/home/geoff/Develop/varnish/varnish-cache/lib/libvmod_purge/.libs:/home/geoff/Develop/varnish/varnish-cache/lib/libvmod_vtc/.libs:/home/geoff/Develop/varnish/varnish-cache/lib/libvmod_blob/.libs:/home/geoff/Develop/varnish/varnish-cache/lib/libvmod_unix/.libs:/home/geoff/Develop/varnish/varnish-cache/lib/libvmod_proxy/.libs",
    "default": "/usr/local/lib/varnish/vmods",
    "description": "Directory (or colon separated list of directories) where VMODs are to be found.",
    "flags": [
    
]
},
  {
    "name": "vsl_space",
    "implemented": true,
    "value": 2097152,
    "units": "bytes",
    "default": "80M",
    "minimum": "1M",
    "maximum": "4G",
    "description": "The amount of space to allocate for the VSL fifo buffer in the VSM memory segment.  If you make this too small, varnish{ncsa|log} etc will not be able to keep up.  Making it too large just costs memory resources.",
    "flags": [
      "must_restart"
]
}
]

param.show -j

[ 2, ["param.show", "-j"], 1538031772.196,
  {
    "name": "accept_filter",
    "implemented": false
  },
  {
    "name": "acceptor_sleep_decay",
    "implemented": true,
    "value": 0.9,
    "default": "0.9",
    "minimum": "0",
    "maximum": "1",
    "description": "If we run out of resources, such as file descriptors or worker threads, the acceptor will sleep between accepts.\nThis parameter (multiplicatively) reduce the sleep duration for each successful accept. (ie: 0.9 = reduce by 10%)",
    "flags": [
      "experimental"
]
},
... and so on for all params
```

Some examples of interest:
```
  {
    "name": "cc_command",
    "implemented": true,
    "value": "exec gcc -g -O2 -Wall -Werror -Wno-error=unused-result -pthread -fpic -shared -Wl,-x -o %o %s",
    "default": "exec gcc -g -O2 -Wall -Werror -Wno-error=unused-result -pthread -fpic -shared -Wl,-x -o %o %s",
    "description": "Command used for compiling the C source code to a dlopen(3) loadable object.  Any occurrence of %s in the string will be replaced with the source file name, and %o will be replaced with the output file name.",
    "flags": [
      "must_reload"
]
},

  {
    "name": "debug",
    "implemented": true,
    "value": "+vtc_mode",
    "default": "none",
    "description": "Enable/Disable various kinds of debugging.\n\tnone\tDisable all debugging\n\nUse +/- prefix to set/reset individual bits:\n\treq_state\tVSL Request state engine\n\tworkspace\tVSL Workspace operations\n\twaiter\tVSL Waiter internals\n\twaitinglist\tVSL Waitinglist events\n\tsyncvsl\tMake VSL synchronous\n\thashedge\tEdge cases in Hash\n\tvclrel\tRapid VCL release\n\tlurker\tVSL Ban lurker\n\tesi_chop\tChop ESI fetch to bits\n\tflush_head\tFlush after http1 head\n\tvtc_mode\tVarnishtest Mode\n\twitness\tEmit WITNESS lock records\n\tvsm_keep\tKeep the VSM file on restart\n\tdrop_pools\tDrop thread pools (testing)\n\tslow_acceptor\tSlow down Acceptor\n\th2_nocheck\tDisable various H2 checks\n\tvmod_so_keep\tKeep copied VMOD libraries\n\tprocessors\tFetch/Deliver processors\n\tprotocol\tProtocol debugging",
    "flags": [
    
]
},

  {
    "name": "feature",
    "implemented": true,
    "value": "none",
    "default": "none",
    "description": "Enable/Disable various minor features.\n\tnone\tDisable all features.\n\nUse +/- prefix to enable/disable individual feature:\n\tshort_panic\tShort panic message.\n\twait_silo\tWait for persistent silo.\n\tno_coredump\tNo coredumps.\n\tesi_ignore_https\tTreat HTTPS as HTTP in ESI:includes\n\tesi_disable_xml_check\tDon't check of body looks like XML\n\tesi_ignore_other_elements\tIgnore non-esi XML-elements\n\tesi_remove_bom\tRemove UTF-8 BOM\n\thttps_scheme\tAlso split https URIs\n\thttp2\tSupport HTTP/2 protocol\n\thttp_date_postel\tRelax parsing of timestamps in HTTP headers",
    "flags": [
    
]
},

  {
    "name": "vsl_mask",
    "implemented": true,
    "value": "-ObjProtocol,-ObjStatus,-ObjReason,-ObjHeader,-VCL_trace,-WorkThread,-Hash,-VfpAcct,-H2RxHdr,-H2RxBody,-H2TxHdr,-H2TxBody",
    "default": "-ObjProtocol,-ObjStatus,-ObjReason,-ObjHeader,-VCL_trace,-WorkThread,-Hash,-VfpAcct,-H2RxHdr,-H2RxBody,-H2TxHdr,-H2TxBody",
    "description": "Mask individual VSL messages from being logged.\n\tdefault\tSet default value\n\nUse +/- prefix in front of VSL tag name, to mask/unmask individual VSL messages.",
    "flags": [
    
]
},
```

Unknown param:
```
param.show -j fofofofo
 106
[ 2, ["param.show", "-j", "fofofofo"], 1538031772.284,

]
Unknown parameter "fofofofo".
```

slimhazard commented:
```
ban.list -j

[ 2, ["ban.list", "-j"], 1538031782.176,
  {
    "time": 1538031782.088210,
    "refs": 0,
    "completed": false,
    "spec": "obj.status == 200"
},
  {
    "time": 1538031782.044229,
    "refs": 0,
    "completed": false,
    "spec": "obj.status == 201"
},
  {
    "time": 1538031781.823607,
    "refs": 1,
    "completed": true,
    "spec": ""
}
]
```

With lurker debugging turned on:
```
ban.list -j
 200
[ 2, ["ban.list", "-j"], 1538031804.596,
  {
    "time": 1538031801.464151,
    "refs": 0,
    "completed": true,
    "spec": "",
    "req_tests": false,
    "obj_tests": true,
    "pointer": "0x7fb384e1c440",
    "objcores": [
    
]
},
  {
    "time": 1538031801.420290,
    "refs": 0,
    "completed": true,
    "spec": "",
    "req_tests": false,
    "obj_tests": true,
    "pointer": "0x7fb384e1c3f0",
    "objcores": [
    
]
},
  {
    "time": 1538031801.376203,
    "refs": 0,
    "completed": true,
    "spec": "",
    "req_tests": false,
    "obj_tests": true,
    "pointer": "0x7fb384e1c3a0",
    "objcores": [
    
]
},
  {
    "time": 1538031801.332138,
    "refs": 0,
    "completed": true,
    "spec": "",
    "req_tests": false,
    "obj_tests": true,
    "pointer": "0x7fb384e1c350",
    "objcores": [
    
]
},
  {
    "time": 1538031801.288226,
    "refs": 0,
    "completed": true,
    "spec": "",
    "req_tests": false,
    "obj_tests": true,
    "pointer": "0x7fb384e1c300",
    "objcores": [
    
]
},
  {
    "time": 1538031795.828147,
    "refs": 0,
    "completed": true,
    "spec": "",
    "req_tests": false,
    "obj_tests": true,
    "pointer": "0x7fb384e1c2b0",
    "objcores": [
    
]
}
]
```
The ban list content changes rapidly, bans are often completed during a vtc test by the time the CLI command gets the chance to run -- this is why, for example, the "objcores" arrays were all empty in the last example.

slimhazard commented:
```
storage.list -j

[ 2, ["storage.list", "-j"], 1538031760.340,
  {
    "name": "s0",
    "storage": "malloc"
},
  {
    "name": "Transient",
    "storage": "malloc"
}
]
```

slimhazard commented:
```
panic.show -j
 200
[ 2, ["panic.show", "-j"], 1538031797.812,
  "Panic at: Thu, 27 Sep 2018 07:03:17 GMT\nWrong turn at cache/cache_main.c:284:\nSignal 11 (Segmentation fault) received at 0x563c941553e0 si_code 2\nversion = varnish-trunk revision 408eb93f26da1f009c72497484e5d6c7900fbcb0, vrt api = 8.0\nident = Linux,4.9.0-8-amd64,x86_64,-jnone,-sdefault,-sdefault,-hcritbit,epoll\nnow = 344253.692624 (mono), 1538031797.724627 (real)\nBacktrace:\n  0x563c9413f9c7: varnishd(+0x4b9c7) [0x563c9413f9c7]\n  0x563c941a5bb0: varnishd(VAS_Fail+0x40) [0x563c941a5bb0]\n  0x563c9413b478: varnishd(+0x47478) [0x563c9413b478]\n  0x7f317c9b40c0: /lib/x86_64-linux-gnu/libpthread.so.0(+0x110c0) [0x7f317c9b40c0]\n  0x7f31745fb63a: vcl_vcl1.1538031797.366053343/vgc.so(VGC_function_vcl_recv+0x1a) [0x7f31745fb63a]\n  0x563c94151bc2: varnishd(+0x5dbc2) [0x563c94151bc2]\n  0x563c9415480a: varnishd(VCL_recv_method+0x5a) [0x563c9415480a]\n  0x563c941447b5: varnishd(CNT_Request+0xa15) [0x563c941447b5]\n  0x563c9416a46a: varnishd(+0x7646a) [0x563c9416a46a]\n  0x563c9415fe3c: varnishd(+0x6be3c) [0x563c9415fe3c]\nthread = (cache-worker)\nthr.req = 0x7f3172440660 {\n  vxid = 1001, transport = HTTP/1 {\n    state = HTTP1::Proc\n  }\n  step = R_STP_RECV,\n  req_body = R_BODY_NONE,\n  restarts = 0, esi_level = 0,\n  sp = 0x7f317220d620 {\n    fd = 22, vxid = 1000,\n    t_open = 1538031797.724547,\n    t_idle = 1538031797.724547,\n    ws = 0x7f317220d660 {\n      id = \\\"ses\\\",\n      {s, f, r, e} = {0x7f317220d698, +96, (nil), +352},\n    },\n    transport = HTTP/1 {\n      state = HTTP1::Proc\n    }\n    client = 127.0.0.1 38984 127.0.0.1:46325,\n  },\n  worker = 0x7f317db30c00 {\n    ws = 0x7f317db30ca8 {\n      id = \\\"wrk\\\",\n      {s, f, r, e} = {0x7f317db303b0, +0, (nil), +2040},\n    },\n    VCL::method = inside RECV,\n    VCL::return = 0x0,\n    VCL::methods = {RECV},\n  },\n  ws = 0x7f31724407b0 {\n    id = \\\"req\\\",\n    {s, f, r, e} = {0x7f31724426e8, +72, (nil), +57168},\n  },\n  http_conn = 0x7f3172442688 {\n    fd = 22 (@0x7f317220d644),\n    doclose = NULL,\n    ws = 0x7f31724407b0 {\n      [Already dumped, see above]\n    },\n    {rxbuf_b, rxbuf_e} = {0x7f31724426e8, 0x7f317244270b},\n    {pipeline_b, pipeline_e} = {(nil), (nil)},\n    content_length = -1,\n    body_status = none,\n    first_byte_timeout = 0.000000,\n    between_bytes_timeout = 0.000000,\n  },\n  http[req] = 0x7f3172440850 {\n    ws = 0x7f31724407b0 {\n      [Already dumped, see above]\n    },\n    hdrs {\n      \\\"GET\\\",\n      \\\"/\\\",\n      \\\"HTTP/1.1\\\",\n      \\\"Host: 127.0.0.1\\\",\n      \\\"X-Forwarded-For: 127.0.0.1\\\",\n    },\n  },\n  vcl = {\n    name = \\\"vcl1\\\",\n    busy = 1,\n    discard = 0,\n    state = auto,\n    temp = warm,\n    conf = {\n      srcname = {\n        \\\"<vcl.inline>\\\",\n        \\\"Builtin\\\",\n      },\n    },\n  },\n  vmods = {\n  },\n  flags = {\n  },\n  privs = 0x7f3172440838 {\n  },\n},\nthr.busyobj = (nil) {\n},\n\n"
]
```
This just converts the panic string into an unstructured JSON string. Conceivably the panic string could be converted to richly structured JSON object, but that would quite a bit more work and probably requires new implementations of everything in child_panic.c. Since the purpose is error reporting, I think it's just as well if clients of the JSON CLI just unescape the JSON string.

rezan commented:
Why a top level array and not an object? With an array, your going to have a huge mess if you ever decide to change the format in the future. Objects with descriptive names a data labels is the universal standard in 2018, no?

bsdphk commented:
@rezan, the top level of JSON responses are there exactly to make versioning easier.

The first member of the array is a version number which describes the layout of the actual response

The second member is the CLI command which elicited the response

The third is a timestamp

The fourth is the actual response, which can be array/dict/string or whatever is most useful.

slimhazard commented:
From bugwash on 2018-10-01:
* There is never JSON output on an error status (so revise the ``param.show`` case above)
* Work in @nigoroll 's comments (and IMO code duplication can be reduced in some other places as well)
* Add a "JSON" doc section to varnish-cli(7), among other things:
  * describe the top-level structure of the array
  * note no JSON for errors

slimhazard commented on bin/varnishd/cache/cache_ban.c:
```diff
@@ -769,9 +769,77 @@ ccf_ban_list(struct cli *cli, const char * const *av, void *priv)
+		VCLI_Out(cli, "\"completed\": %s,\n",
+			 b->flags & BANS_FLAG_COMPLETED ? "true" : "false");
+		VCLI_Out(cli, "\"spec\": \"");
+		ban_render(cli, b->spec);
```
@nigoroll none of the test outputs required it. The only thing I see in ban_render() that might require quoting is the output of the ban argument: ``VCLI_Out(cli, "%s", bt.arg2);`` (currently line 719 in cache_ban.c). Could that potentially contain something that needs to be quoted? (You know bans better than I do.) If you can write up an example it would help a lot, then I'd add it to the tests and write up the code.

Generally I'm trying to avoid rewriting existing code like ban_render() where possible. But if the quoting can be limited to that one line, it won't be too obtrusive

slimhazard replied:
Ok just tried a test using a ban expression in which the arg is quoted (and will be adding it to c00019.vtc):
```
# Ban expression with quoting
varnish v1 -cliok {ban req.url ~ "BAR"}
varnish v1 -cliok "ban.list"
varnish v1 -clijson "ban.list -j"
```

The ban expression with quoting is permitted (note: on my reading, vcl(7) implies that it would have been an error, but it evidently isn't), but the quotation marks are removed in the ``ban.list`` output.

From this I infer that no part of ``ban.list`` output requires quoting in the JSON rendering.

@nigoroll I'm still convinced that you know bans better than I do, so I don't mind if you can prove me wrong.

slimhazard commented on bin/varnishd/mgt/mgt_param.c:
```diff
@@ -335,6 +335,150 @@ mcf_param_show(struct cli *cli, const char * const *av, void *priv)
+		flags = 0;
+		VCLI_Out(cli, "\"flags\": [\n");
+		VSB_indent(cli->sb, 2);
+		if (pp->flags & OBJ_STICKY) {
```
We don't have a header with the macros in a table -- the flag definitions are inlined in the declaration of ``struct parspec`` in ``mgt_param.h``. What I'll do here is define and expand macros right in ``mgt_param.c``.

slimhazard commented:
Changes implemented as described in the last comment, rebased to current master and force-pushed.

* ``param.show -j``, and all other commands supporting JSON, do not generate JSON output for an error status -- the output is the same as the version without ``-j``.
* The implementation of ``ban.list`` has been refactored so that the code executing the mutex/refcounting before and after the output is not duplicated.
* But quoting is not implemented to render ``ban.list -j``, since I believe it is never necessary, for the reasons given in the exchange with @nigoroll  further up. Happy to fix that if I got that wrong.
* The implementation of ``param.show`` uses a macro to reduce the code duplication to generate the array of flags.
* Added a "JSON" section to varnish-cli(7), documenting the top-level structure of JSON output, and the fact that there is no JSON output for an error response.

slimhazard commented:
Changes made, rebased to current master and force-pushed.

* For the ``ban.list`` case, just check if av[2] is "-j". (Now there's no multiline loop after all.

* @nigoroll showed me an example for which quoting is necessary in the output of ``ban.list`` -- do ``varnishadm ban 'obj.http.Host ~ \"Foo\"'`` and then ``varnishadm ban.list``. So ban_render() now uses VCLI_Quote() to render the ban expression in the JSON case, VCLI_Out() otherwise. The example has been added to c00019.vtc.
